### PR TITLE
Disallow unknown JSON keys in standard-json

### LINKED ISSUE #4902: Disallow unsupported json keys in standard-json
The standard-json compiler currently just picks the json fields it knows about and ignores the rest. We should think about whether it makes sense to disallow anything it does not know about.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #4902

### CODE REVIEW

File: libsolidity/interface/StandardCompiler.cpp [hunk 1]
```diff
@@ -227,13 +229,62 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+	auto const& members = _input.getMemberNames();
+	for (auto const& member: members)
+		if (!binary_search(_keys.begin(), _keys.end(), member))
+			return formatFatalError("JSONError", "Unknown key " + member);
+	return boost::none;
+}
+
+boost::optional<Json::Value> StandardCompiler::checkRootKeys(Json::Value const& _input)
+{
+	vector<string> keys{"auxiliaryInput", "language", "settings", "sources"};
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: static?

File: libsolidity/interface/StandardCompiler.cpp [hunk 2]
```diff
@@ -227,13 +229,62 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 
 }
 
+boost::optional<Json::Value> StandardCompiler::checkKeys(Json::Value const& _input, vector<string>& _keys)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
boost::optional<Json::Value> StandardCompiler::checkKeys(Json::Value const& _input, vector<string> const& _keys)
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 3]
```diff
@@ -227,13 +229,62 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+{
+	sort(_keys.begin(), _keys.end());
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Do we really want to sort them for each invocation?

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -227,13 +229,62 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+	auto const& members = _input.getMemberNames();
+	for (auto const& member: members)
+		if (!binary_search(_keys.begin(), _keys.end(), member))
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Why not just use a statically declared `std::set`?

File: libsolidity/interface/StandardCompiler.h [hunk 5]
```diff
@@ -55,6 +57,21 @@ class StandardCompiler: boost::noncopyable
 private:
 	Json::Value compileInternal(Json::Value const& _input);
 
+	/// Checks whether _input has at least one member that is not in _keys.
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.h]: These could be in an anonymous namespace.

**Comment by leonardoalt** [comment]
@chriseth Updated

File: test/cmdlineTests.sh [hunk 6]
```diff
@@ -113,15 +113,20 @@ printTask "Testing unknown options..."
...
+    local stdout_expected="${4}"
+    local exit_code_expected="${5}"
+    local stderr_expected="${6}"
     local stdout_path=`mktemp`
     local stderr_path=`mktemp`
     if [[ "$exit_code_expected" = "" ]]; then exit_code_expected="0"; fi
 
     set +e
-    "$SOLC" "${filename}" ${solc_args} 1>$stdout_path 2>$stderr_path
+	if [[ "$solc_stdin" = "" ]]; then
```
Comment (chriseth) [inline test/cmdlineTests.sh]: indentation

File: libsolidity/interface/StandardCompiler.cpp [hunk 7]
```diff
@@ -225,6 +228,51 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 	return output;
 }
 
+boost::optional<Json::Value> checkKeys(Json::Value const& _input, set<string> const& _keys)
+{
+	auto const& members = _input.getMemberNames();
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: No need for a local variable.

File: libsolidity/interface/StandardCompiler.cpp [hunk 8]
```diff
@@ -225,6 +228,51 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+	for (auto const& member: members)
+		if (_keys.find(member) == _keys.end())
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: count?

File: libsolidity/interface/StandardCompiler.cpp [hunk 9]
```diff
@@ -225,6 +228,51 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+			return formatFatalError("JSONError", "Unknown key " + member);
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Perhaps quote `member`?

File: Changelog.md [hunk 10]
```diff
@@ -9,6 +9,7 @@ Compiler Features:
  * Code Generator: Use codecopy for string constants more aggressively.
  * Code Generator: Use binary search for dispatch function if more efficient. The size/speed tradeoff can be tuned using ``--optimize-runs``.
  * SMTChecker: Support mathematical and cryptographic functions in an uninterpreted way.
+ * Standard JSON Interface: disallow unknown keys.
```
Comment (erak) [inline Changelog.md]:
```suggestion
 * Commandline interface: Disallow unknown keys in standard JSON input.
```

**Review by erak — COMMENTED** [review]
Could you move the ``*.json``, ``*.stdout`` & ``*.exit`` into their own folder? We have a few of them already in root directory. Otherwise this looks fine :)

**Comment by chriseth** [comment]
I agree, we should clean up those commandline tests and have one directory per test, but I think this can be done in another PR.

**Comment by chriseth** [comment]
Created https://github.com/ethereum/solidity/issues/5640

File: Changelog.md [hunk 11]
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * Inline Assembly: Improve error messages around invalid function argument count.
  * Code Generator: Use codecopy for string constants more aggressively.
  * Code Generator: Use binary search for dispatch function if more efficient. The size/speed tradeoff can be tuned using ``--optimize-runs``.
+ * Commandline Interface: Disallow unknown keys in standard JSON input.
```
Comment (chriseth) [inline Changelog.md]: Strictly, this is not just about the commandline interface. Perhaps `standard-json interface`?

Comment (chriseth) [reply Changelog.md]: Or just "Compiler Interface"?

Comment (erak) [reply Changelog.md]: Sorry, that I made you change that in the first place, @leonardoalt!

File: libsolidity/interface/StandardCompiler.cpp [hunk 12]
```diff
@@ -225,6 +228,50 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 	return output;
 }
 
+boost::optional<Json::Value> checkKeys(Json::Value const& _input, set<string> const& _keys)
+{
```
Comment (Marenz) [inline libsolidity/interface/StandardCompiler.cpp:232]: I would also check at this point whether `_input` is an object

Comment (Marenz) [reply libsolidity/interface/StandardCompiler.cpp:232]: but only error out when it also exists and is not an object ;)

Comment (Marenz) [reply libsolidity/interface/StandardCompiler.cpp:232]: Actually, I can just add in on my PR which is based on this one

Comment (leonardoalt) [reply libsolidity/interface/StandardCompiler.cpp:232]: Ok!

File: Changelog.md (hunk 11 above)
Comment (chriseth) [reply Changelog.md]: Ah, just saw that now. Sorry! :)

**Review by chriseth — APPROVED** [review]

**Pull request merged**